### PR TITLE
Update PHP-CS-Fixer to the v3.60 

### PR DESCRIPTION
**Pull request by viktorkulahin**
PHP-CS-Fixer has been updated to the v3.60 
There are a lot of changes relative to deprecations and name changes because of this update.
Code style changes.

The update is not complete. Some more work has to be done. Currently, all tests are passed. But some of the fixers are off. For example, because the Parser doesn't work correctly with some new features in php8. It can't parse classes with class member declarations in the constructor. This is only one example of existing problems.

But in general, it could be a good start for the Paysera PHP CS Fisex 3.0.0.

### CODE REVIEW

**Review by matasarei — APPROVED** [review]

File: defaults/php-cs-fixer-risky.php [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<?php
+declare(strict_types=1);
```
Comment (mSprunskas) [inline defaults/php-cs-fixer-risky.php:2]: We follow PSR-12 style https://github.com/paysera/php-style-guide this means linebreak is needed after php opening tag. Please fix this in other places too.

File: src/Config/PayseraConventionsConfig.php [hunk 2]
```diff
@@ -111,9 +110,10 @@ public function __construct()
             new BracesFixer(),
             new DefaultValuesInConstructorFixer(),
         ]));
+        $this->migrationModeRules = null;
```
Comment (mSprunskas) [inline src/Config/PayseraConventionsConfig.php:113]: why not empty array?

**Review by mSprunskas — CHANGES_REQUESTED** [review]
Please rebase, I'd like to review changes in proper IDEA, but rebase is needed first, thanks.

File: defaults/php-cs-fixer-risky.php (hunk 1 above)
Comment (viktorkulahin) [reply defaults/php-cs-fixer-risky.php:3]: Done

File: src/Config/PayseraConventionsConfig.php (hunk 2 above)
Comment (viktorkulahin) [reply src/Config/PayseraConventionsConfig.php]: Fixed

File: src/Fixer/PhpBasic/CodeStyle/DocBlockWhitespaceFixer.php [hunk 3]
```diff
@@ -62,14 +66,15 @@ protected function applyFix(SplFileInfo $file, Tokens $tokens)
...
+                            $matches[1],
                         );
 
                         $lines[$key] = implode($matches);
                     }
                 }
 
-                $token->setContent(implode($lines));
+                $tokens[$index] = new Token([$token->getId(), implode($lines)]);
+//                $token->setContent(implode($lines));
```
Comment (mSprunskas) [inline src/Fixer/PhpBasic/CodeStyle/DocBlockWhitespaceFixer.php:77]: No need to leave commented out code, prev code version is available in git history

Comment (viktorkulahin) [reply src/Fixer/PhpBasic/CodeStyle/DocBlockWhitespaceFixer.php]: Fixed

File: src/Util/InheritanceHelper.php [hunk 4]
```diff
@@ -10,60 +11,57 @@
...
         } catch (ReflectionException $exception) {
             return false;
         }
 
-        foreach ($reflection->getInterfaces() as $interface) {
-            $methods = $interface->getMethods(ReflectionProperty::IS_PUBLIC | ReflectionProperty::IS_PROTECTED);
-            foreach ($methods as $method) {
-                if ($method->getName() === $methodName) {
-                    return true;
+        if ($reflection !== null) {
```
Comment (mSprunskas) [inline src/Util/InheritanceHelper.php:22]: Good opportunity to use guard clause
```
if ($reflection === null) {
    return false;
}
// ...
```
In other places too.

File: tests/AbstractFixerTestCase.php [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+    }
+
+    protected function lintSource(string $source): ?string
+    {
+        try {
+            $this->linter->lintSource($source)
+                ->check()
+            ;
+        } catch (Exception $e) {
+            return $e->getMessage() . "\n\nSource:\n{$source}";
```
Comment (mSprunskas) [inline tests/AbstractFixerTestCase.php:554]: use sprintf, see https://github.com/paysera/php-style-guide?tab=readme-ov-file#double-quotes

File: tests/AbstractFixerTestCase.php [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+    ];
+
+    abstract protected function createFixer();
+
+    final public function testIsRisky()
+    {
+        if ($this->fixer->isRisky()) {
+            self::assertValidDescription($this->fixer->getName(), 'risky description', $this->fixer->getDefinition()->getRiskyDescription());
+        } else {
+            self::assertNull($this->fixer->getDefinition()->getRiskyDescription(), \sprintf('[%s] Fixer is not risky so no description of it expected.', $this->fixer->getName()));
```
Comment (mSprunskas) [inline tests/AbstractFixerTestCase.php:69]: `\` qualifier in `\sprintf` is redundant

File: tests/AbstractFixerTestCase.php [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+
+        if ($this->fixer instanceof HeaderCommentFixer) {
+            $this->fixer->configure(['header' => 'a']);
+        }
+
+        static::assertNull(
+            $this->fixer->fix(new SplFileInfo(__FILE__), $emptyTokens),
+            sprintf('Return type for ::fix with empty tokens of "%s" is invalid.', $this->fixer->getName()),
+        );
+        static::assertFalse($emptyTokens->isChanged());
```
Comment (mSprunskas) [inline tests/AbstractFixerTestCase.php:470]: duplicate assertion

File: tests/AbstractFixerTestCase.php [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+
+        static::assertNull(
+            $this->fixer->fix(new SplFileInfo(__FILE__), $tokens),
```
Comment (mSprunskas) [inline tests/AbstractFixerTestCase.php:473]: assert is wrong, return type of `\PhpCsFixer\Fixer\FixerInterface::fix` is void, better assert that exception was not raised or that some changes occurred.

File: tests/AbstractFixerTestCase.php [hunk 9]
```diff
@@ -0,0 +1,658 @@
...
+        $reflection = new ReflectionMethod($this->fixer, 'isRisky');
+
+        // If fixer is not risky then the method `isRisky` from `AbstractFixer` must be used
+        self::assertSame(
+            !$this->fixer->isRisky(),
+            AbstractFixer::class === $reflection->getDeclaringClass()->getName()
+        );
+    }
+
+    final public function testFixerDefinitions()
```
Comment (mSprunskas) [inline tests/AbstractFixerTestCase.php:85]: Refactoring is needed, method length is almost 180+ lines...

I guess this is a result of not following https://github.com/paysera/php-style-guide?tab=readme-ov-file#composition-over-inheritance

**Comment by mSprunskas** [comment]
Changelog entry is needed.

File: src/Parser/Entity/SeparatedItemList.php [hunk 10]
```diff
@@ -41,15 +43,12 @@ public function getSeparatorItems(): Generator
...
-     * @return ItemInterface|null
-     */
-    public function getSeparatorAfterContents()
+    public function getSeparatorAfterContents(): ?ItemInterface
     {
         return $this->getFirstItemOfType(TypedItem::TYPE_SEPARATOR_AFTER_CONTENTS);
     }
 
-    public function areSeparatorsSplitIntoSeveralLines()
+    public function areSeparatorsSplitIntoSeveralLines(): bool
```
[the remark this answers is not in the document]
Comment (viktorkulahin) [reply src/Parser/Entity/SeparatedItemList.php:51]: Removed

File: src/Util/InheritanceHelper.php (hunk 4 above)
Comment (viktorkulahin) [reply src/Util/InheritanceHelper.php:22]: Fixed

File: tests/AbstractFixerTestCase.php (hunk 5 above)
Comment (viktorkulahin) [reply tests/AbstractFixerTestCase.php:554]: This is old code. It is fixed anyway.

File: tests/AbstractFixerTestCase.php (hunk 6 above)
Comment (viktorkulahin) [reply tests/AbstractFixerTestCase.php:69]: Fixed

File: tests/AbstractFixerTestCase.php (hunk 7 above)
Comment (viktorkulahin) [reply tests/AbstractFixerTestCase.php:470]: fixed

File: tests/AbstractFixerTestCase.php (hunk 9 above)
Comment (viktorkulahin) [reply tests/AbstractFixerTestCase.php:85]: Yes, refactoring is needed. This method was taken from the old php-cs-fixer library, from AbstractFixerTestCase. Let's fix it after the merge.

File: tests/AbstractFixerTestCase.php (hunk 8 above)
Comment (viktorkulahin) [reply tests/AbstractFixerTestCase.php:479]: Will be fixed after refactoring

**Review by mSprunskas — APPROVED** [review]
In order to keep things simpler lets implement composition pattern in tests in next PR

**Pull request merged**